src/common/melody_generator.py

```diff
@@ -0,0 +1,134 @@
+import tensorflow.keras as keras
+import json
+import numpy as np
+import music21 as m21
+
+
+class MelodyGenerator:
+
+    # TODO: Proper dependency injection (is that even possible in Python?)
+    def __init__(self, model_path: str, music_helper, file_helper, mapping_path: str, sequence_length: int):
+        self.model = keras.models.load_model(model_path)
+        self._music_helper = music_helper
+        self._file_helper = file_helper
+        self._start_symbols = ["/"] * sequence_length # This acts as our song delimtter 
+
+        # TODO: move to file_helper
+        with open(mapping_path, "r") as fp:
+            self._mappings = json.load(fp)
+
+    
+    def generate_melody(self, seed, num_steps, max_seq_len, temperature):
+        """ 
+        * seed - what we want to pass to the network and the network countinues that seed
+        * num_steps - How far we want the network to predict
+        * max_seq_len - How many steps of the seeds do we want to consider for the network (the seed will grow large), 
+        so we pass the seed as a sliding window
+        """
+
+        # create seed with start symbols
+        # it is passed to us as a str, let us turn it into a list
+        seed = seed.split()
+        melody = seed
+        seed = self._start_symbols + seed
+
+        # map seed to int
+        seed = [self._mappings[symbol] for symbol in seed]
+
+        for _ in range(num_steps):
+            # limit the seed to max_seq_len
+            seed = seed[-max_seq_len:] 
+
+            # ont-hot encode the seed
+            onehot_seed = keras.utils.to_categorical(seed, num_classes=len(self._mappings))
+
+            # Keras expects 3-dimensions for predict so we use numpy to add that
+            # (1, max_seq_len, len(mappings))
+            onehot_seed = onehot_seed[np.newaxis, ...]
+
+            # inference stage, this will result in a distribution of probabilities, but we just want the first/most likely one
+            # i.e: [0.]
+            probabilities = self.model.predict(onehot_seed)[0]
+            output_int = self._sample_with_temperature(probabilities, temperature)
+
+            # update seed
+            seed.append(output_int)
+
+            # map int to our encoding
+            output_symbol = [k for k, v in self._mappings.items() if v == output_int][0]
+
+            # check whether we're at the end of a melody
+            if output_symbol == "/":
+                break
+
+            # update the melody
+            melody.append(output_symbol)
+
+        return melody
+
+
+    def _sample_with_temperature(self, probabilities, temperature=0.7):
+        """ 
+        The temperature modifies the probability distribution
+        the lower the temperature, the more deterministic the sampling is
+        if the temperature is 1 that is just default settings
+        higher temperature, more exploritive model
```

Comment:
Adding the datatypes for the parameters may also be helpful to others adding to this.